### PR TITLE
Deprecate `PPUPPredFailure`

### PR DESCRIPTION
**Pull request by lehins**
# Description

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/cardano-ledger-core/src/Cardano/Ledger/Core/Era.hs [hunk 1]
```diff
@@ -20,7 +22,8 @@ module Cardano.Ledger.Core.Era (
   -- ** Rules
   EraRule,
   EraRuleFailure,
-  VoidRuleFailure,
```
Comment (teodanciu) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Core/Era.hs:23]: Should we mention the removal of `VoidRuleFailure` in CHANGELOG ?   (or its replacement with `VoidEraRule`)

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxo.hs [hunk 2]
```diff
@@ -27,11 +33,31 @@ instance InjectRuleFailure "UTXO" AlonzoUtxoPredFailure (ConwayEra c) where
...
+  Allegra.OutsideValidityIntervalUTxO vi slotNo -> OutsideValidityIntervalUTxO vi slotNo
+  Allegra.MaxTxSizeUTxO x y -> MaxTxSizeUTxO x y
+  Allegra.InputSetEmptyUTxO -> InputSetEmptyUTxO
+  Allegra.FeeTooSmallUTxO c1 c2 -> FeeTooSmallUTxO c1 c2
+  Allegra.ValueNotConservedUTxO vc vp -> ValueNotConservedUTxO vc vp
+  Allegra.WrongNetwork x y -> WrongNetwork x y
+  Allegra.WrongNetworkWithdrawal x y -> WrongNetworkWithdrawal x y
+  Allegra.OutputTooSmallUTxO x -> OutputTooSmallUTxO x
+  Allegra.UpdateFailure x -> absurdEraRule @"PPUP" @era x
+  Allegra.OutputBootAddrAttrsTooBig xs -> OutputTooBigUTxO (map (0,0,) xs)
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxo.hs]: I wonder  why it's not `OutputBootAddrAttrsTooBig xs` here, but I can see that this is how it is in `allegraToAlonzoUtxoPredFailure`, so it's probably correct.

**Review by teodanciu — APPROVED** [review]
Looks good to me

File: libs/cardano-ledger-core/src/Cardano/Ledger/Core/Era.hs (hunk 1 above)
Comment (lehins) [reply libs/cardano-ledger-core/src/Cardano/Ledger/Core/Era.hs:23]: Yes, `VoidEraRule` is a replacement. I'll adjust the changelog. thanks for noticing

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxo.hs (hunk 2 above)
Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxo.hs]: That is an interesting observation. I have not looked into why it was that way. I only copy pasted the `allegraToAlonzoUtxoPredFailure` and adjusted the part that was relevant for Conway.

It looks like a mistake in injection that was introduced here: https://github.com/IntersectMBO/cardano-ledger/commit/47d8a35bf31fdd5c1bb5c20b40abd8303884e13f#diff-a9a408a2190f9a7d639380d4faf3d0d0178c9b96d2195c42d8bf7fdd505faa3aR566

I'll fix it for Conway and we can clean it up for older eras later.

Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxo.hs]: Done

**Pull request merged**